### PR TITLE
Bug fixes

### PR DESCRIPTION
**Pull request by Crisspl**
## Description
- [x] PRIORITY BUG (https://github.com/Devsh-Graphics-Programming/Nabla/pull/38/commits/b7b12efe18005fc6cecc0568f9aa7b98496e33e9) PERFORMANCE CRITICAL
Descriptor Set Binding Managment jest zjebany... odpal sobie commit przed i ten commit raytraced AO example to zobaczysz
- [x] RGB19E7 nie dziala jako storage format (ustaw emissive na vec3(0.0,1.0,0.0) w raygen.glsl a color na 0.0,0.0) to zobaczysz flicker/cykl z full zielony do czarnego i spowrotem na startupie
- [x] Example 22 crashuje sie przy quitcie, costam odnosnie asset cache albo cos (warte uwagi)
- [x] inny concern (design bug) 
```
    // TODO: @Crisspl find a way to stop the user from such insanity as moving from the bundle's dynamic array
        //m_visibilityBufferFillPipelineLayoutGPU = std::move(m_driver->getGPUObjectsFromAssets<ICPUPipelineLayout>(&m_visibilityBufferFillPipelineLayoutCPU,&m_visibilityBufferFillPipelineLayoutCPU+1u)->operator[](0));
```
ah w sumie tam to ok...
ale chodzi o to zebym nie mogl zrobic std::move z jakiego cache
- [x] IOpenGLPipeline array stride calculation smells fishy
- [x] jeszcze bug, jak odpalisz daily_pt.xml to jest crash (shader fails compile) with
```
nbl/builtin/glsl/material_compiler/common.glsl:143: error: 'op_isBXDF' : no matching overloaded function found
nbl/builtin/glsl/material_compiler/common.glsl:143: error: 'return' : type does not match, or is not convertible to, the function's return type
```

## TODO list:
unchecked things above

### CODE REVIEW

File: examples_tests/22.RaytracedAO/dirty_source/ExtraCrap.cpp [hunk 1]
```diff
@@ -21,7 +21,7 @@ constexpr uint32_t kOptiXPixelSize = sizeof(uint16_t)*3u;
 core::smart_refctd_ptr<ICPUSpecializedShader> specializedShaderFromFile(IAssetManager* assetManager, const char* path)
 {
 	auto bundle = assetManager->getAsset(path, {});
-	return core::move_and_static_cast<ICPUSpecializedShader>(*bundle.getContents().begin());
```
Comment (devshgraphicsprogramming) [inline examples_tests/22.RaytracedAO/dirty_source/ExtraCrap.cpp:24]: is this why the sample crashed at quit?

Comment (Crisspl) [reply examples_tests/22.RaytracedAO/dirty_source/ExtraCrap.cpp:24]: yep, you've moved from a cached bundle

Comment (devshgraphicsprogramming) [reply examples_tests/22.RaytracedAO/dirty_source/ExtraCrap.cpp:24]: yeah, we need to make a way on the `smart_refctd_ptr/refctd_dynamic_array` which is being used in the `SBundle` to forbid that insanity

Comment (devshgraphicsprogramming) [reply examples_tests/22.RaytracedAO/dirty_source/ExtraCrap.cpp:24]: this is what the `inny concern (design bug)` should be about (not moving from GPU conversion result... I got my wires crossed)

Comment (Crisspl) [reply examples_tests/22.RaytracedAO/dirty_source/ExtraCrap.cpp:24]: yea, i figured

File: src/nbl/video/COpenGLRenderpassIndependentPipeline.h [hunk 2]
```diff
@@ -36,7 +36,8 @@ class COpenGLRenderpassIndependentPipeline final : public IGPURenderpassIndepend
             _vertexInputParams, _blendParams, _primAsmParams, _rasterParams
             ),
             IOpenGLPipeline(_ctxCount, _ctxID, _GLnames, _binaries),
-            m_stagePresenceMask(0u)
+            m_stagePresenceMask(0u),
+            m_baseInstanceUniformIDs(core::make_refctd_dynamic_array<decltype(m_baseInstanceUniformIDs)>(_ctxCount))
```
Comment (devshgraphicsprogramming) [inline src/nbl/video/COpenGLRenderpassIndependentPipeline.h:40]: this hurts... can you not slap on an extra 4 bytes to the uniform variables state? and use it out of there?

File: src/nbl/asset/CShaderIntrospector.cpp [hunk 3]
```diff
@@ -698,16 +698,18 @@ static void introspectStructType(spirv_cross::Compiler& _comp, impl::SShaderMemo
                 member.count_specID = sc->id;
             }
         }
-		else if (mtype.basetype != spirv_cross::SPIRType::Struct) // might have to ignore a few more types than structs
-			member.arrayStride = core::max(0x1u<<core::findMSB(member.size),16u); // @Crisspl  NOT TRUE FOR MATRICES!
+        else if (mtype.columns > 1u)
+            member.arrayStride = member.size;
+		else if (mtype.basetype != spirv_cross::SPIRType::Struct)
+			member.arrayStride = core::max(0x1u<<core::findMSB(member.size),16u);
```
Comment (devshgraphicsprogramming) [inline src/nbl/asset/CShaderIntrospector.cpp:704]: explain your reasoning?

Before all types except for structs got their strides aligned up to PoT, the only problem was that non square matrices (or even a `mat2`) might have a size bigger than 16 bytes and their array stride could be out of whack.

Now you've disabled this stride determination for everything but matrices (no vectors) and the same issue as above remains for matrices.

File: source/Nabla/COpenGLDriver.cpp [hunk 4]
```diff
@@ -1997,13 +2002,14 @@ void COpenGLDriver::drawIndexedIndirect(const asset::SBufferBinding<IGPUBuffer>
...
 {
+    const COpenGLPipelineLayout* prevLayout = effectivelyBoundDescriptors.layout.get();
     //bind new descriptor sets
     int32_t compatibilityLimit = 0u;
-    if (_prevLayout && _currentLayout)
-        compatibilityLimit = _prevLayout->isCompatibleUpToSet(IGPUPipelineLayout::DESCRIPTOR_SET_COUNT-1u, _currentLayout)+1u;
-	if (!_prevLayout && !_currentLayout)
+    if (prevLayout && _currentLayout)
+        compatibilityLimit = prevLayout->isCompatibleUpToSet(IGPUPipelineLayout::DESCRIPTOR_SET_COUNT-1u, _currentLayout)+1u;
+	if (!prevLayout && !_currentLayout)
```
Comment (devshgraphicsprogramming) [inline source/Nabla/COpenGLDriver.cpp:2012]: now you've got a bug here.

Comment (devshgraphicsprogramming) [reply source/Nabla/COpenGLDriver.cpp:2012]: actually fine

File: src/nbl/video/IOpenGLPipeline.h [hunk 5]
```diff
@@ -73,15 +73,18 @@ class IOpenGLPipeline
...
+                // in case of non-array types, m.arrayStride is irrelevant
+                // we should compute it though, so that we dont have to branch in the loop 
+                if (!m.isArray())
                 {
-                    uint32_t arrayStride1 = 0u;
+                    // 1N for scalar types, 2N for gvec2, 4N for gvec3 and gvec4
+                    // N==sizeof(float)
                     if (is_scalar_or_vec())
-                        arrayStride1 = (m.mtxRowCnt==1u) ? m.size : core::roundUpToPoT(m.mtxRowCnt)*4u;
+                        arrayStride = (m.mtxRowCnt==1u) ? m.size : core::roundUpToPoT(m.mtxRowCnt)*sizeof(float);
```
Comment (devshgraphicsprogramming) [inline src/nbl/video/IOpenGLPipeline.h:84]: make a note that if we ever start supporting doubles in push constants, this needs to be adapted

File: include/nbl/builtin/glsl/format/encode.glsl [hunk 6]
```diff
@@ -9,9 +9,14 @@ uvec2 nbl_glsl_encodeRGB19E7(in vec3 col)
...
-
-	const uvec3 mantissas = uvec3(clamped*exp2(nbl_glsl_RGB19E7_MANTISSA_BITS-shared_exp) + 0.5);
+	int shared_exp = clamp(f32_exp,-nbl_glsl_RGB19E7_EXP_BIAS-1,nbl_glsl_MAX_RGB19E7_EXP) + 1;
+
+	float scale = exp2(nbl_glsl_RGB19E7_MANTISSA_BITS - shared_exp);
+	const uint maxm = uint(maxrgb*scale + 0.5);
+	const bool need = (maxm == nbl_glsl_MAX_RGB19E7_MANTISSA_VALUES);
+	scale = need ? 0.5*scale : scale;
+	shared_exp = need ? (shared_exp+1) : shared_exp;
+	const uvec3 mantissas = uvec3(clamped*scale + 0.5);
```
Comment (devshgraphicsprogramming) [inline include/nbl/builtin/glsl/format/encode.glsl:19]: doesn't  `decode` need to be changed accordingly?

Comment (Crisspl) [reply include/nbl/builtin/glsl/format/encode.glsl:19]: nah, decode just reads what encode have written

Comment (devshgraphicsprogramming) [reply include/nbl/builtin/glsl/format/encode.glsl:19]: if something was wrong with encode, then probably something wrong with decode as well

Also, do you mind optimizing?

**Pull request merged**